src/rules/switchDefaultRule.ts

```diff
@@ -36,16 +54,66 @@ export class Rule extends Lint.Rules.AbstractRule {
     public static FAILURE_STRING = "Switch statement should include a 'default' case";
 
     public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
-        return this.applyWithFunction(sourceFile, walk);
+        return this.applyWithFunction(sourceFile, walk, {
+            allowExhaustive: false, // This option requires TypeChecker
+        });
+    }
+
+    public applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): Lint.RuleFailure[] {
+        return this.applyWithFunction(
+            sourceFile,
+            walk,
+            {
+                allowExhaustive: this.ruleArguments.indexOf(OPTION_ALLOW_EXHAUSTIVE) !== -1,
+            },
+            program.getTypeChecker(),
+        );
     }
 }
 
-function walk(ctx: Lint.WalkContext<void>) {
+function walk(ctx: Lint.WalkContext<Options>, checker?: ts.TypeChecker) {
     return ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
-        if (node.kind === ts.SyntaxKind.SwitchStatement &&
-            !(node as ts.SwitchStatement).caseBlock.clauses.some(isDefaultClause)) {
-            ctx.addFailureAtNode(node, Rule.FAILURE_STRING);
+        if (
+            ts.isSwitchStatement(node)
+            && !node.caseBlock.clauses.some(isDefaultClause)
+        ) {
+            const failure = !checker || !ctx.options.allowExhaustive
+                ? `${Rule.FAILURE_STRING}`
+                : checkForUncoveredValues(
+                    checker.getTypeAtLocation(node.expression),
+                    node.caseBlock,
+                );
+
+            if (failure !== undefined) {
+                ctx.addFailureAtNode(node, failure);
+            }
         }
         return ts.forEachChild(node, cb);
     });
 }
+
+function checkForUncoveredValues(exprType: ts.Type, caseBlock: ts.CaseBlock): string | void {
+    if (!isTypeFlagSet(exprType, ts.TypeFlags.Union)) {
+        return Rule.FAILURE_STRING;
+    }
+
+    for (const type of (exprType as ts.UnionType).types) {
+        if (!isTypeFlagSet(type, ts.TypeFlags.StringLiteral)) {
```

Comment:
We should not limit this to string literal types. There are a few more to consider: number literal types, boolean literal types, null, undefined.